pkg/e2e/machinehealthcheck/machinehealthcheck.go

```diff
@@ -0,0 +1,230 @@
+package machinehealthcheck
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/golang/glog"
+	e2e "github.com/openshift/cluster-api-actuator-pkg/pkg/e2e/framework"
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	healthcheckingv1alpha1 "github.com/openshift/machine-api-operator/pkg/apis/healthchecking/v1alpha1"
+	"github.com/openshift/machine-api-operator/pkg/util/conditions"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck controller", func() {
+	var client runtimeclient.Client
+	var numberOfReadyWorkers int
+	var workerNode *corev1.Node
+	var workerMachine *mapiv1beta1.Machine
+
+	stopKubeletAndValidateMachineDeletion := func(workerNodeName *corev1.Node, workerMachine *mapiv1beta1.Machine, timeout time.Duration) {
+		By(fmt.Sprintf("Stopping kubelet service on the node %s", workerNode.Name))
+		err := e2e.StopKubelet(workerNode.Name)
+		Expect(err).ToNot(HaveOccurred())
+
+		By(fmt.Sprintf("Validating that node %s has 'NotReady' condition", workerNode.Name))
+		waitForNodeUnhealthyCondition(workerNode.Name)
+
+		By(fmt.Sprintf("Validating that machine %s is deleted", workerMachine.Name))
+		machine := &mapiv1beta1.Machine{}
+		key := types.NamespacedName{
+			Namespace: workerMachine.Namespace,
+			Name:      workerMachine.Name,
+		}
+		Eventually(func() bool {
+			err := client.Get(context.TODO(), key, machine)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					return true
+				}
+			}
+			return false
+		}, timeout, 5*time.Second).Should(BeTrue())
+	}
+
+	BeforeEach(func() {
+		var err error
+		client, err = e2e.LoadClient()
+		Expect(err).ToNot(HaveOccurred())
+
+		listOptions := runtimeclient.ListOptions{
+			Namespace: e2e.TestContext.MachineApiNamespace,
+		}
+		listOptions.SetLabelSelector(fmt.Sprintf("%s=", e2e.NodeWorkerLabel))
+		workers := &corev1.NodeList{}
+		err = client.List(context.TODO(), &listOptions, workers)
+		Expect(err).ToNot(HaveOccurred())
+
+		numberOfReadyWorkers = 0
+		workerNode = nil
+		for i, w := range workers.Items {
+			readyCond := conditions.GetNodeCondition(&w, corev1.NodeReady)
+			if readyCond.Status == corev1.ConditionTrue {
+				numberOfReadyWorkers++
+				if workerNode == nil {
+					workerNode = &workers.Items[i]
+					glog.V(2).Infof("Worker node %s", workerNode.Name)
+				}
+			}
+		}
+		Expect(workerNode).ToNot(BeNil())
+
+		listOptions = runtimeclient.ListOptions{
+			Namespace: e2e.TestContext.MachineApiNamespace,
+		}
+		machineList := &mapiv1beta1.MachineList{}
+		err = client.List(context.TODO(), &listOptions, machineList)
+		Expect(err).ToNot(HaveOccurred())
+
+		for i, m := range machineList.Items {
```

Comment:
This deserves its own function and put it into utils. There's a library meant to be reusable in infra/utils. So then ginkgo e2e user stories are abstracted and easy to read with out caring about details